File: Go/PollingSystem2-go/repositories/poll_repository.go
package repositories

import (
	"fmt"
	"sync"
	"time"

	"lld-practice/pollingsystem2-go/models"
)

type PollRepository struct {
	mu     sync.RWMutex
	polls  []*models.Poll
	nextID int
}

func NewPollRepository() *PollRepository {
	return &PollRepository{polls: []*models.Poll{}, nextID: 1}
}

func (r *PollRepository) NextID() int {
	r.mu.Lock()
	defer r.mu.Unlock()
	id := r.nextID
	r.nextID++
	return id
}

func (r *PollRepository) Add(poll *models.Poll) error {
	if poll == nil {
		return fmt.Errorf("invalid poll")
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	for _, p := range r.polls {
		if p.ID == poll.ID {
			return fmt.Errorf("poll already exists!")
		}
	}
	r.polls = append(r.polls, poll)
	return nil
}

func (r *PollRepository) GetByID(id int) *models.Poll {
	r.mu.RLock()
	defer r.mu.RUnlock()
	for _, poll := range r.polls {
		if poll.ID == id {
			return poll
		}
	}
	return nil
}

func (r *PollRepository) Update(poll *models.Poll) error {
	if poll == nil {
		return fmt.Errorf("invalid poll!")
	}
	r.mu.Lock()
	defer r.mu.Unlock()
	for i, p := range r.polls {
		if p.ID == poll.ID {
			r.polls[i] = poll
			return nil
		}
	}
	return fmt.Errorf("poll not found!")
}

func (r *PollRepository) GetActive(now time.Time) []*models.Poll {
	r.mu.RLock()
	defer r.mu.RUnlock()
	var active []*models.Poll
	for _, poll := range r.polls {
		if !poll.IsCompleted(now) {
			active = append(active, poll)
		}
	}
	return active
}

func (r *PollRepository) GetCompletedByCreator(creatorID int, now time.Time) []*models.Poll {
	r.mu.RLock()
	defer r.mu.RUnlock()
	var completed []*models.Poll
	for _, poll := range r.polls {
		if poll.CreatedBy == creatorID && poll.IsCompleted(now) {
			completed = append(completed, poll)
		}
	}
	return completed
}

File: JavaScript/PollingSystem2/PollingService/repositories/PollRepository.js
const { Poll } = require('../models/Poll');

/** In-memory stand-in for a poll repository. */
class PollRepository {
  static polls = [];
  static nextId = 1;

  static createId() {
    return this.nextId++;
  }

  static add(poll) {
    if (!(poll instanceof Poll)) {
      throw new Error('Invalid poll');
    }

    if (this.polls.some((p) => p.id === poll.id)) {
      throw new Error('Poll already exists!');
    }

    this.polls.push(poll);
  }

  static getById(id) {
    return this.polls.find((p) => p.id === id);
  }

  static update(poll) {
    if (!(poll instanceof Poll)) {
      throw new Error('Invalid poll!');
    }

    const index = this.polls.findIndex((p) => p.id === poll.id);
    this.polls[index] = poll;
  }

  static getActive(now = Date.now()) {
    return this.polls.filter((poll) => !poll.isCompleted(now));
  }

  static getCompletedByCreator(creatorId, now = Date.now()) {
    return this.polls.filter(
      (poll) => poll.createdBy === creatorId && poll.isCompleted(now),
    );
  }
}

module.exports = { PollRepository };
